jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImplTest.java

```diff
@@ -807,4 +810,68 @@ public void testUserTasksInCaseAdStakeholder() {
             }
         }
     }
+    
+    @Test
+    public void testTransitionBetweenStagesInCase() {
+        // use user name who is part of the case roles assignment
+        // so (s)he will be authorized to access case instance
+        identityProvider.setName("john");
+        Map<String, OrganizationalEntity> roleAssignments = new HashMap<>();
+        roleAssignments.put("owner", new UserImpl("john"));
+        
+        assertNotNull(deploymentService);        
+        DeploymentUnit deploymentUnit = new KModuleDeploymentUnit(GROUP_ID, ARTIFACT_ID, VERSION);
+        
+        deploymentService.deploy(deploymentUnit);
+        units.add(deploymentUnit);
+        Map<String, Object> data = new HashMap<>();
+        CaseFileInstance caseFile = caseService.newCaseFileInstance(deploymentUnit.getIdentifier(), TWO_STAGES_CASE_P_ID, data, roleAssignments);
+        String caseId = caseService.startCase(deploymentUnit.getIdentifier(), TWO_STAGES_CASE_P_ID, caseFile);
+        assertNotNull(caseId);
+        try {
+            
+            Collection<CaseStageInstance> stage = caseRuntimeDataService.getCaseInstanceStages(caseId, true, new QueryContext(0, 1));
+            assertNotNull(stage);
+            assertEquals(1, stage.size());
+            assertEquals("Stage One", stage.iterator().next().getName());
+            assertEquals(StageStatus.Active, stage.iterator().next().getStatus());
+            
+            Collection<AdHocFragment> adhocTasks = caseRuntimeDataService.getAdHocFragmentsForCase(caseId);
+            assertNotNull(adhocTasks);
+            assertEquals(1, adhocTasks.size());
+            assertEquals("Task 1", adhocTasks.iterator().next().getName());
+            
+            Collection<NodeInstanceDesc> activeNodes = caseRuntimeDataService.getActiveNodesForCase(caseId, new QueryContext(0, 10));
+            assertNotNull(activeNodes);
+            assertEquals(1, activeNodes.size());
+            assertEquals("Stage One", activeNodes.iterator().next().getName());
+            
+            caseService.addDataToCaseFile(caseId, "customData", "nextStagePlease");
+            
+            stage = caseRuntimeDataService.getCaseInstanceStages(caseId, true, new QueryContext(0, 1));
```

Comment:
its actually fails because paging is applied on db and thus returns first found row. That is because NodeInstanceLog has two rows per each node instance - entry and exit thus filtering on db is expensive. So taking into consideration stages there won't be many in given case - like not 10000 etc then better is to do filtering in the code. The fix for this is here https://github.com/droolsjbpm/jbpm/pull/741 including your test case so I will close this PR.